### DISCUSSION TITLE
Nested middlewares

### DISCUSSION THREAD

**Discussion by predaytor** [vercel/next.js#38728] [Help]
I have a core middleware with Content-Security-Policy applied to all routes and a `/projects` route, for example, where I do some logic for that particular route. There are many other routes, now need to import all their specific middleware code into one file? It looks very counterintuitive. What's the purpose of a new exported config if our middleware will eventually take care of all the routes?

It looks like Fresh (Deno) took the right path by reviewing the past Next.js solution
https://fresh.deno.dev/docs/concepts/middleware

How can I organize this code now? I would like to use this old way :( 

`_middleware.ts`
```ts
import { NextResponse } from 'next/server';

export function middlware() {
	const response = NextResponse.next();

	...
	response.headers.set('X-DNS-Prefetch-Control', 'on');

	return response;
}
```
`projects/_middleware.ts`
```ts
import { NextRequest, NextResponse } from 'next/server';

export function middleware(request: NextRequest) {
	return NextResponse.redirect(new URL('/', request.url));
}

export const config = {
	matcher: '/projects',
};
```

**Comment by GabenGar** [comment]
Just add more `if` statements.

**Reply by predaytor** [reply]
A project has many routes, so putting every import specific to one route into a single middleware file is not great idea. I just don't understand the design decision the team made. The old approach was correct in my opinion.

**Reply by FlurryNight** [reply]
> A project has many routes, so putting every import specific to one route into a single middleware file is not great idea. I just don't understand the design decision the team made. The old approach was correct in my opinion.

Exactly, we should be able to attach middlewares to routes like in Slim PHP framework, putting everything in on file may not work or be a lot of mess with if statements all over the place.

**Reply by predaytor** [reply]
> > A project has many routes, so putting every import specific to one route into a single middleware file is not great idea. I just don't understand the design decision the team made. The old approach was correct in my opinion.
> 
> Exactly, we should be able to attach middlewares to routes like in Slim PHP framework, putting everything in on file may not work or be a lot of mess with if statements all over the place.

I'm thinking of switching to remix based on express.js and deploying on fly.io with docker, nginx, redis. The project is completely static with CMS integration (with previews) and page pagination. I think the benefits of using SSG are not that valuable in a real application (http caching with revalidation is enough). So, I will have complete freedom of architecture and no build step on every update (or preview request).

**Reply by FlurryNight** [reply]
Okay,just looked at the link for fresh by deno that u sent, that's is what we should be able to do here, but

**Reply by FlurryNight** [reply]
You also have this [remix-middleware](https://github.com/neurosnap/remix-middleware), althougth never tested it
